### PR TITLE
4641 ruby shared lib

### LINKED ISSUE #4641: Python Measures: Factor Ruby related source out of cli and into a shared library
This is probably the highest point of risk in the introduction of Python Measures. The idea is to put the ruby related embedded files, the ruby engine, and all "the trimmings" into a shared library, therefore making the cli no longer fully self contained. Currently this content is statically linked and embedded directly into the eopnstudio command line executable program.

### PR DESCRIPTION
**Pull request by kbenne**
Pull request overview
---------------------

- Fix #4641

This PR moves the Ruby related parts of OpenStudio into a shared library that is dynamically loaded by the CLI.

There is significant project cleanup and reorganization. 

## Pull Request Author

 - [ ] Everything builds
 - [ ] Ruby bindings can be imported in ruby
     - [ ] Windows
     - [ ] Linux
     - [ ] Mac
- [ ] CLI works on all platforms
     - [ ] Windows
     - [ ] Linux
     - [ ] Mac
 - [ ] NREL/OpenStudio-resources testing
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [x] If change to an IDD file, add the label `IDDChange`
 - [x] If breaking existing API, add the label `APIChange`
 - [x] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [x] Labeling is ok
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

**Comment by kbenne** [comment]
@tijcolem This might not be ready, but it is at least time to get CI looking at it. In particular it needs a cross platform examination, which CI should help with.

**Comment by jmarrec** [comment]
I'm going to cherry pick the first commits that delete a bunch of cruft into a separate branch and merge that to develop so we can clean up the diff here.

File: ruby/README.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    * The target `rubyinterpreter-declarations` is an interface library that merely adds Ruby to the include path. This is used by the standalone ruby bindings.
+    * The target `rubyinterpreter` adds Ruby to the include paths, but also links to the Ruby library. This is used by the OpenStudio Ruby engine.
+
+* `bindings`: This contains an interface to initialize the OpenStudio API bindings for Ruby. (e.g. `init_openstudio_model`)
+The `module` and `engine` components consume this.
+
+* `module`: This is the conventional OpenStudio Ruby bindings, for loading into a standalone Ruby interpreter.
+
+* `engine`: This is a combination of the OpenStudio Ruby bindings, the Ruby interpreter, and embedded files
+to be used by the OpenStudio cli to run Ruby based workflows.
```
Comment (jmarrec) [inline ruby/README.md:17]: Where does src/scriptengine play into this?

File: src/scriptengine/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# Script Engine is a base for scripting in OpenStudio
+# See implementations in ruby/engine and python/engine
+
+add_library(scriptengine INTERFACE)
+target_include_directories(scriptengine INTERFACE ${CMAKE_CURRENT_SOURCE_DIR})
```
Comment (jmarrec) [inline src/scriptengine/CMakeLists.txt:5]: I'm confused by this subdirectory, and it's location at src/scriptengine. Right now it's only used by the ruby/engine target.

File: ruby/CMakeLists.txt [hunk 3]
```diff
@@ -1,191 +1,8 @@
...
-
-endif()
+# OpenStudio Ruby components
+
+add_subdirectory(interpreter)
+add_subdirectory(bindings)
+add_subdirectory(module)
+add_subdirectory(engine)
+# TODO: Enable this
+#add_subdirectory(test)
```
Comment (jmarrec) [inline ruby/CMakeLists.txt:8]: - [ ]  TODO

File: src/cli/main.cpp [hunk 4]
```diff
@@ -582,30 +19,36 @@ int main(int argc, char* argv[]) {
...
-    return 0;
+  if ((argc > 1) && (std::string_view(argv[1]) == "experimental")) {
+    CLI::App app{"openstudio"};
+
+    const auto experimentalApp = app.add_subcommand("experimental");
+    [[maybe_unused]] const auto runCommand = experimentalApp->add_subcommand("run");
+
+    CLI11_PARSE(app, argc, argv);
+  } else {
+    result = rubyCLI(argc, argv);
```
Comment (jmarrec) [inline src/cli/main.cpp:47]: I have the start of a CLI11 cpp parser at https://github.com/jmarrec/OSCLI11/tree/main/src

File: src/utilities/core/DynamicLibraryWindows.hpp [hunk 5]
```diff
@@ -0,0 +1,103 @@
+#ifndef sdynamiclibrary_windows_hpp_INCLUDED
+#define sdynamiclibrary_windows_hpp_INCLUDED
+
+#include "Path.hpp"
+#include <fmt/format.h>
+#include <string>
+#include <filesystem>
+#include <Windows.h>
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/utilities/core/DynamicLibraryWindows.hpp:8]: I forgot I actually enabled that in CMakeLists.txt: https://github.com/NREL/OpenStudio/blame/d5e18658c14b658a8010595ea4096054e56246eb/CMakeLists.txt#L382

https://github.com/NREL/OpenStudio/blob/d5e18658c14b658a8010595ea4096054e56246eb/CMakeLists.txt#L382

**Comment by jmarrec** [comment]
macOS and Ubuntu 18.04 have a too old cmake.

```
CMake Error at ruby/interpreter/CMakeLists.txt:20 (add_library):
  add_library INTERFACE library requires no source arguments.
```

https://cmake.org/cmake/help/latest/command/add_library.html#interface-libraries

> New in version 3.15: An interface library can have [PUBLIC_HEADER](https://cmake.org/cmake/help/latest/prop_tgt/PUBLIC_HEADER.html#prop_tgt:PUBLIC_HEADER) and [PRIVATE_HEADER](https://cmake.org/cmake/help/latest/prop_tgt/PRIVATE_HEADER.html#prop_tgt:PRIVATE_HEADER) properties. The headers specified by those properties can be installed using the [install(TARGETS)](https://cmake.org/cmake/help/latest/command/install.html#command:install) command.

> **New in version 3.19: An interface library target may be created with source files:**

File: src/scriptengine/CMakeLists.txt (hunk 2 above)
Comment (kbenne) [reply src/scriptengine/CMakeLists.txt:5]: Soon Python will use it also.

File: ruby/CMakeLists.txt (hunk 3 above)
Comment (kbenne) [reply ruby/CMakeLists.txt:8]: Actually these tests are now covered. https://github.com/NREL/OpenStudio/blob/4641-ruby-shared-lib/src/cli/CMakeLists.txt#L59

I believe the TODO comment should be removed.

Comment (jmarrec) [reply ruby/CMakeLists.txt:8]: Not the same thing. One runs the tests with system ruby (-I<...>/openstudio.so), the clitest run with the cli.
I thought I had pushed a commit that re-enables these (and deletes the TODO) but I apparently didn't.

Comment (jmarrec) [reply ruby/CMakeLists.txt:8]: rubytests can be run via the CLI, the reverse isn't true (eg: embedded gems)

Comment (kbenne) [reply ruby/CMakeLists.txt:8]: I see.

**Comment by tijcolem** [comment]
@jmarrec @kbenne  Windows CI finally ran and completed.  Looks like one unit test failure and looks like related to ruby.  Everything looks good. 

```
0/11 Test #3406: RubyTest-UUID_Test-uuid_big_set ......................***Failed    3.81 sec
MiniTest::Unit::TestCase is now Minitest::Test. From D:/jenkins/openstudio/develop/ruby/test/./UUID_Test.rb:34:in `<main>'
E, [2022-09-21T19:38:48.905375 #7232] ERROR -- : Unknown arguments ["'D:/jenkins/openstudio/develop/ruby/test/./UUID_BigSet_Aux.rb'", "'1000'"] found
Usage: openstudio [options] <command> [<args>]
    -h, --help                       Print this help.
        --verbose                    Print the full log to STDOUT
    -I, --include DIR                Add additional directory to add to front of Ruby $LOAD_PATH (may be used more than once)
    -e, --execute CMD                Execute one line of script (may be used more than once). Returns after executing commands.
        --gem_path DIR               Add additional directory to add to front of GEM_PATH environment variable (may be used more than once)
        --gem_home DIR               Set GEM_HOME environment variable
        --bundle GEMFILE             Use bundler for GEMFILE
        --bundle_path BUNDLE_PATH    Use bundler installed gems in BUNDLE_PATH
        --bundle_without WITHOUT_GROUPS
                                     Space separated list of groups for bundler to exclude in WITHOUT_GROUPS.  Surround multiple groups with quotes like "test development".
```

**Comment by jmarrec** [comment]
Superseded by #4688

**Review by tijcolem — APPROVED** [review]
@jmarrec @kbenne What impressive work by the both of you! This will be such an awesome first step in supporting Python measures.

**Pull request merged**